pkg/portForward/portForward.go

```diff
@@ -0,0 +1,148 @@
+package portForward
+
+import (
+	"fmt"
+	"io"
+	"reflect"
+
+	"github.com/devfile/api/v2/pkg/apis/workspaces/v1alpha2"
+	"github.com/devfile/library/pkg/devfile/parser"
+	parsercommon "github.com/devfile/library/pkg/devfile/parser/data/v2/common"
+
+	"github.com/redhat-developer/odo/pkg/kclient"
+	"github.com/redhat-developer/odo/pkg/libdevfile"
+	"github.com/redhat-developer/odo/pkg/log"
+	"github.com/redhat-developer/odo/pkg/state"
+	"github.com/redhat-developer/odo/pkg/util"
+)
+
+var _ Client = (*PFClient)(nil)
+
+type PFClient struct {
+	kubernetesClient kclient.ClientInterface
+	stateClient      state.Client
+
+	appliedEndpoints map[string][]int
+
+	// stopChan on which to write to stop the port forwarding
+	stopChan chan struct{}
+	// finishedChan is written when the port forwarding is finished
+	finishedChan chan struct{}
+}
+
+func NewPFClient(kubernetesClient kclient.ClientInterface, stateClient state.Client) *PFClient {
+	return &PFClient{
+		kubernetesClient: kubernetesClient,
+		stateClient:      stateClient,
+	}
+}
+
+func (o *PFClient) StartPortForwarding(
+	devFileObj parser.DevfileObj,
+	randomPorts bool,
+	errOut io.Writer,
+) error {
+
+	// get the endpoint/port information for containers in devfile and setup port-forwarding
+	containers, err := devFileObj.Data.GetComponents(parsercommon.DevfileOptions{
+		ComponentOptions: parsercommon.ComponentOptions{ComponentType: v1alpha2.ContainerComponentType},
+	})
+	if err != nil {
+		return err
+	}
+	ceMapping := libdevfile.GetContainerEndpointMapping(containers)
+	if o.stopChan != nil && reflect.DeepEqual(ceMapping, o.appliedEndpoints) {
+		return nil
```

Comment:
We are first getting what are the endpoints defined in the devfile, with ceMapping. 
If nothing changed (DeepEqual) since the last time (o.stopChan != nil) we forwared the ports, we return immediately.

From another point of view, every time we start the forwarding, we have a non-nil stopChan, and every time we stop them, we set stopChan to nil, so stopChan represents when the forwarding is running: we continue only if the forwarding is not started or if the endpoints have changed in the Devfile

Reply:
> every time we start the forwarding, we have a non-nil stopChan

Don't we have a `nil` channel when we start port forwarding? I see it being set using `make` a few lines after this check.

> we continue only if the forwarding is not started or if the endpoints have changed in the Devfile

👍🏾